src/esp32/src/flash.c

```diff
@@ -4,16 +4,148 @@
  * SPDX-License-Identifier: Apache-2.0 OR MIT
  */
 
-#include "flash.h"
+#include <target/flash.h>
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include <log.h>
+#include <private/rom.h>
+#include <private/soc_utils.h>
+
+#include "soc/gpio_reg.h"
+#include "soc/spi_reg.h"
+
+#define SPI_FLASH_CMD_RDID 0x9F
+#define ESP_ROM_SPIFLASH_BUSY_FLAG BIT(0)
+
+#define PERIPHS_SPI_FLASH_CMD SPI_CMD_REG(1)
+#define PERIPHS_SPI_FLASH_STATUS SPI_RD_STATUS_REG(1)
+#define PERIPHS_SPI_FLASH_USRREG SPI_USER_REG(1)
+#define PERIPHS_SPI_FLASH_USRREG2 SPI_USER2_REG(1)
+#define PERIPHS_SPI_FLASH_C0 SPI_W0_REG(1)
+#define PERIPHS_SPI_MOSI_DLEN_REG SPI_MOSI_DLEN_REG(1)
+#define PERIPHS_SPI_MISO_DLEN_REG SPI_MISO_DLEN_REG(1)
+#define SPI_USR2_DLEN_SHIFT SPI_USR_COMMAND_BITLEN_S
+
+/* ROM */
+
+extern uint32_t ets_efuse_get_spiconfig(void);
+extern void esp_rom_spiflash_attach(uint32_t ishspi, bool legacy);
+extern esp_rom_spiflash_chip_t g_rom_flashchip;
+extern uint8_t g_rom_spiflash_dummy_len_plus[];
 
 void stub_target_flash_init(void *state)
 {
+    STUB_LOG_TRACE();
     (void)state;
-    // TODO: Implement
+    uint32_t spiconfig = ets_efuse_get_spiconfig();
+
+    uint32_t strapping = REG_READ(GPIO_STRAP_REG);
+    /*  If GPIO1 (U0TXD) is pulled low and flash pin configuration is not set in efuse, assume
+     * HSPI flash mode (same as normal boot) */
+    if (spiconfig == 0 && (strapping & 0x1c) == 0x08) {
```

Comment:
I know we are copying this lines somewhere else, but it would be nice to know what is this magic numbers meaning? 0x1c, 0x8.

Reply:
Yep, strap stuff is something that will still need putting order. It was just moved from the original stub

Reply:
updated in target_flash.c, please let me know if it is not clear